### PR TITLE
executor,metrics: add a metric for observing execution phases

### PR DESCRIPTION
**Pull request by zyguan**
Signed-off-by: zyguan <zhongyangguan@gmail.com>

### What problem does this PR solve?

Issue Number: ref #34106

Problem Summary:

The execution duraition is the main part of the database time, however it cannot be broken down (in wall time) by current metrics.

### What is changed and how it works?

The execution process can be described as the following.

The main parts are `build`, `open`, `next`, `lock` and `commit` and we may retry `build -> open -> next [-> lock]` when a pessimistic lock error returned. Thus these phases (except for `commit`) can be further split into two parts: 
1. trying to lock keys (but failed)
2. the final iteration of the retry loop

By observing durations of these phases, we can identify some typical issues easily, eg:

1. too much time spent on building executor (typically caused by waiting tso)

2. lock contention

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: executor/adapter.go [hunk 1]
```diff
@@ -896,8 +930,119 @@ var (
...
+	execNextLocking  = metrics.ExecPhaseDuration.WithLabelValues("next:locking", "0")
+	execLockLocking  = metrics.ExecPhaseDuration.WithLabelValues("lock:locking", "0")
+	execBuildFinal   = metrics.ExecPhaseDuration.WithLabelValues("build:final", "0")
+	execOpenFinal    = metrics.ExecPhaseDuration.WithLabelValues("open:final", "0")
+	execNextFinal    = metrics.ExecPhaseDuration.WithLabelValues("next:final", "0")
+	execLockFinal    = metrics.ExecPhaseDuration.WithLabelValues("lock:final", "0")
+
+	execCommitPrewrite     = metrics.ExecPhaseDuration.WithLabelValues("commit:prewrite", "0")
+	execCommitCommit       = metrics.ExecPhaseDuration.WithLabelValues("commit:commit", "0")
+	execCommitWaitCommitTS = metrics.ExecPhaseDuration.WithLabelValues("commit:wait:commit-ts", "0")
```
Comment (zyguan) [inline executor/adapter.go:945]: TODO: add `commit:wait:latest-ts` after https://github.com/tikv/client-go/pull/538 gets merged.

File: executor/adapter.go [hunk 2]
```diff
@@ -900,8 +934,127 @@ var (
...
+	execCommitWaitLatestTS = metrics.ExecPhaseDuration.WithLabelValues("commit:wait:latest-ts", "0")
+	execCommitWaitLatch    = metrics.ExecPhaseDuration.WithLabelValues("commit:wait:local-latch", "0")
+	execCommitWaitBinlog   = metrics.ExecPhaseDuration.WithLabelValues("commit:wait:prewrite-binlog", "0")
 )
 
+func (a *ExecStmt) observePhaseDurations(internal bool, commitDetails *util.CommitDetails) {
+	if d := a.phaseBuildDurations[0]; d > 0 {
+		if internal {
+			metrics.ExecPhaseDuration.WithLabelValues("build:final", "1").Observe(d.Seconds())
```
Comment (cfzjywxk) [inline executor/adapter.go:958]: Maybe we could abstract the label constants and pre-define the related metrics?

File: executor/adapter.go [hunk 3]
```diff
@@ -216,6 +216,11 @@ type ExecStmt struct {
 	retryCount        uint
 	retryStartTime    time.Time
 
+	phaseBuildDurations [2]time.Duration
```
Comment (cfzjywxk) [inline executor/adapter.go:219]: Better to add some comments for the array definition and the reset function.

File: executor/adapter.go [hunk 4]
```diff
@@ -900,8 +934,127 @@ var (
...
 	totalTiFlashQuerySuccCounter      = metrics.TiFlashQueryTotalCounter.WithLabelValues("", metrics.LblOK)
+
+	execBuildLocking = metrics.ExecPhaseDuration.WithLabelValues("build:locking", "0")
+	execOpenLocking  = metrics.ExecPhaseDuration.WithLabelValues("open:locking", "0")
+	execNextLocking  = metrics.ExecPhaseDuration.WithLabelValues("next:locking", "0")
+	execLockLocking  = metrics.ExecPhaseDuration.WithLabelValues("lock:locking", "0")
+	execBuildFinal   = metrics.ExecPhaseDuration.WithLabelValues("build:final", "0")
+	execOpenFinal    = metrics.ExecPhaseDuration.WithLabelValues("open:final", "0")
+	execNextFinal    = metrics.ExecPhaseDuration.WithLabelValues("next:final", "0")
+	execLockFinal    = metrics.ExecPhaseDuration.WithLabelValues("lock:final", "0")
```
Comment (cfzjywxk) [inline executor/adapter.go:945]: For the `next` operation maybe we need to integrate more with the kv client metrics in the future.

File: executor/adapter.go (hunk 2 above)
Comment (zyguan) [reply executor/adapter.go]: updated

File: executor/adapter.go [hunk 5]
```diff
@@ -724,6 +736,7 @@ func (a *ExecStmt) handlePessimisticDML(ctx context.Context, e Executor) error {
 		ctx = context.WithValue(ctx, util.LockKeysDetailCtxKey, &lockKeyStats)
 		startLocking := time.Now()
 		err = txn.LockKeys(ctx, lockCtx, keys...)
+		a.phaseLockDurations[0] += time.Since(startLocking)
```
Comment (you06) [inline executor/adapter.go:739]: Point get and batch point get will lock keys internally, that duration will not be observed here.

Comment (zyguan) [reply executor/adapter.go:739]: Yes, here the lock phase is only for pessimistic DMLs that have extra keys to lock. The internal lock key duration (of point-get, point-update, etc) is counted in next phase.

File: executor/adapter.go [hunk 6]
```diff
@@ -896,12 +937,130 @@ func FormatSQL(sql string) stringutil.StringerFunc {
...
+	phaseOpenFinal          = "open:final"
+	phaseNextFinal          = "next:final"
+	phaseLockFinal          = "lock:final"
+	phaseCommitPrewrite     = "commit:prewrite"
+	phaseCommitCommit       = "commit:commit"
+	phaseCommitWaitCommitTS = "commit:wait:commit-ts"
+	phaseCommitWaitLatestTS = "commit:wait:latest-ts"
+	phaseCommitWaitLatch    = "commit:wait:local-latch"
+	phaseCommitWaitBinlog   = "commit:wait:prewrite-binlog"
+	phaseWriteResponse      = "write-response"
```
Comment (zyguan) [inline executor/adapter.go:955]: @cfzjywxk @you06 @sticnarf @longfangsong Any suggestion about naming? eg. `build:locking` and `locking:build`, which is better?

**Comment by zyguan** [comment]
There are still some unknown spans. I found diagnosis (stmt summary & top sql, not included in this PR) may cost more time than writing response and there is 5.24s database time (execute:other) I cannot explain currently. Since the total duation (5.24s+1.65s) accounts for only 3.64% database time, I think we can just ignore them for now.

File: executor/adapter.go (hunk 6 above)
Comment (you06) [reply executor/adapter.go:955]: `build:locking` IMO, it's better to put the related metrics together in the alphabetic order.

**Comment by cfzjywxk** [comment]
> diagnosis (stmt summary & top sql, not included in this PR) may cost more time than writing response
Seems the workload does not return much data to the client?

**Comment by cfzjywxk** [comment]
@zyguan 
Usually, the `next` operation may take much time, maybe we need to find a way to integrate it with the kv client durations later.

File: metrics/executor.go [hunk 7]
```diff
@@ -46,4 +46,13 @@ var (
 			Name:      "statement_db_total",
 			Help:      "Counter of StmtNode by Database.",
 		}, []string{LblDb, LblType})
+
+	// ExecPhaseDuration records the duration of each execution phase.
+	ExecPhaseDuration = prometheus.NewSummaryVec(
```
Comment (sticnarf) [inline metrics/executor.go:51]: Can you explain why summary is used here? I have never seen summary type metrics used in TiDB before and histogram is always used....

**Comment by zyguan** [comment]
> @zyguan Usually, the `next` operation may take much time, maybe we need to find a way to integrate it with the kv client durations later. Also after merging this pr the performance-map needs to be updated.

Yes, we need to figure out a way to show these durations in wall time, since kv client / executors may run concurrently.

File: metrics/executor.go (hunk 7 above)
Comment (zyguan) [reply metrics/executor.go:51]: To reduce the size of metrics data (histogram = summary + buckets). IMO, we won't care too much about sth like "what's the p99 latency of next phase". There are too many kinds of executors (as well as their combinations), some may be fast and other may be extremely slow, a higher or lower p99 latency may not provide more info (we do not known about the distribution of each kind of executors). Besides, it's hard to decide buckets here, some phases (like open) take very little time, but phases like lock may cost a few seconds.

**Review by cfzjywxk — APPROVED** [review]
LGTM

**Review by you06 — APPROVED** [review]

**Pull request merged**